mlir-tensorrt/tensorrt/lib/Target/TensorRTEncodingOpInterface/NetworkEncoder.cpp

```diff
@@ -922,6 +922,29 @@ LogicalResult NvInferNetworkEncoder::encodeFunc(FunctionOpInterface func) {
       return failure();
     nvinfer1::ITensor *inputTensor =
         getNetworkDefinition()->addInput(name.c_str(), *dtype, trtShape);
+
+    // setDimensionName must be called immediately after addInput, or TensorRT
+    // will not deduplicate equal dimensions, which leads to perf gaps.
+    auto dimNamesAttr = func.getArgAttrOfType<DictionaryAttr>(
+        arg.getArgNumber(), TensorRTDialect::getDimensionNamesArgAttrName());
+    if (dimNamesAttr) {
+
+      for (NamedAttribute namedAttr : dimNamesAttr) {
+        int32_t key;
+        if (namedAttr.getName().getValue().getAsInteger(10, key)) {
+          return func->emitOpError()
+                 << "dimension name key '" << namedAttr.getName()
+                 << "' is not an integer";
+        }
+
+        if (StringAttr strAttr = namedAttr.getValue().dyn_cast<StringAttr>()) {
```

Comment:
Failure to cast to `StringAttr` should be an error, no?

Also, we should be checking for `key` being out-of-bounds here.

Technically, the correct way to verify all these things are correct is to use  a Discardable Attribute Verifier (https://mlir.llvm.org/docs/DefiningDialects/#discardable-attribute-verification) on the TensorRT dialect. That would ensure any appearance of `tensorrt.dimension_names` could be verified to ensure it conforms to the correct requirements. But since we don't already have that for the other func arg/result attributes we have been associating with the TensorRT dialect, I don't think you have to introduce it in this MR. Until we have that verifier, however, we should do our best to emit errors for out-of-spec uses of the arg attr here.

Reply:
Yes, changed it to emit an error and also added checks for the key being in-bounds.